### PR TITLE
kv: reduce the distsql concurrency if the runaway query action is COOLDOWN

### PR DESCRIPTION
**Pull request by nolouch**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/55029

Problem Summary:

Now the effect of `COOLDOWN` in runaway query is not very significant, and also we meet the big query problem need to reduce the concurrency when executing to smoothly consume the RU,  so combine the requirements. optimize the `COOLDOWN` action by reduce the distsql concurrency.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/store/copr/coprocessor.go [hunk 1]
```diff
@@ -219,6 +220,15 @@ func (c *CopClient) BuildCopIterator(ctx context.Context, req *kv.Request, vars
 		it.concurrency = 1
 	}
 
+	// if the request is triggered cool down by the runaway checker, we need to adjust the concurrency, let the sql run slowly.\
```
Comment (HuSharp) [inline pkg/store/copr/coprocessor.go:223]:
```suggestion
	// if the request is triggered cool down by the runaway checker, we need to adjust the concurrency, let the sql run slowly.
```

**Review by HuSharp — APPROVED** [review]

File: pkg/domain/resourcegroup/runaway.go [hunk 2]
```diff
@@ -525,15 +525,29 @@ func (r *RunawayChecker) BeforeExecutor() error {
...
+func (r *RunawayChecker) CheckAction() rmpb.RunawayAction {
+	if r == nil {
+		return rmpb.RunawayAction_NoneAction
+	}
+	if r.markedByRule.Load() {
+		return r.setting.Action
+	}
+	if r.markedByWatch.Load() {
+		return r.watchAction
+	}
```
Comment (JmPotato) [inline pkg/domain/resourcegroup/runaway.go:538]: Should the query watch have a higher priority?

File: pkg/store/copr/coprocessor.go [hunk 3]
```diff
@@ -219,6 +220,15 @@ func (c *CopClient) BuildCopIterator(ctx context.Context, req *kv.Request, vars
...
+	if len(req.ResourceGroupName) > 0 && req.ResourceGroupName != resourcegroup.DefaultResourceGroupName && req.RunawayChecker != nil {
```
Comment (glorv) [inline pkg/store/copr/coprocessor.go:224]:
```suggestion
	if req.RunawayChecker != nil {
```
Why need to check ResourceGroupName?

**Comment by nolouch** [comment]
PTAL @glorv  @JmPotato

File: pkg/domain/resourcegroup/runaway.go [hunk 4]
```diff
@@ -505,7 +505,7 @@ func (r *RunawayChecker) BeforeExecutor() error {
 			if action == rmpb.RunawayAction_NoneAction && r.setting != nil {
 				action = r.setting.Action
 			}
-			r.markedByWatch = true
+			r.markedByWatch.Store(true)
```
[the remark this answers is not in the document]
Comment (glorv) [reply pkg/domain/resourcegroup/runaway.go:509]: But you only *read* the value in `BeforeCop`, that say, after the first set in `ExecStmt::Exec`, `markedByWatch` is never changed, so there is no need to use atomic.

File: pkg/store/copr/coprocessor.go [hunk 5]
```diff
@@ -219,6 +220,15 @@ func (c *CopClient) BuildCopIterator(ctx context.Context, req *kv.Request, vars
 		it.concurrency = 1
 	}
 
+	// if the request is triggered cool down by the runaway checker, we need to adjust the concurrency, let the sql run slowly.
+	if req.RunawayChecker != nil {
+		action := req.RunawayChecker.CheckAction()
+		if action == rmpb.RunawayAction_CoolDown {
```
Comment (JmPotato) [inline pkg/store/copr/coprocessor.go:226]:
```suggestion
	if req.RunawayChecker != nil && req.RunawayChecker.CheckAction() == rmpb.RunawayAction_CoolDown {
```

**Review by JmPotato — APPROVED** [review]
The rest LGTM.

**Comment by nolouch** [comment]
ptal @glorv @tangenta

**Review by glorv — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

**Pull request merged**